dev/com.ibm.ws.security.authorization.jacc.ejb/src/com/ibm/ws/security/authorization/jacc/ejb/impl/EJBJaccServiceImpl.java

```diff
@@ -0,0 +1,313 @@
+    public void propagateEJBRoles(BeanMetaData bmd) {
+        propagateEJBRoles(esp, bmd.j2eeName.getApplication(), bmd.j2eeName.getModule(), bmd.enterpriseBeanName, bmd.ivRoleLinkMap,
+                          JaccUtil.convertMethodInfoList(JaccUtil.mergeMethodInfos(bmd)));
+    }
+
+    protected void propagateEJBRoles(EJBSecurityPropagator esp,
+                                     String applicationName,
+                                     String moduleName,
+                                     String beanName,
+                                     Map<String, String> roleLinkMap,
+                                     Map<RoleInfo, List<MethodInfo>> methodMap) {
+        JaccService jaccService = jaccServiceRef.getService();
+        if (jaccService != null) {
+            esp.propagateEJBRoles(jaccService.getContextId(applicationName, moduleName), applicationName, beanName, roleLinkMap, methodMap,
+                                  jaccService.getPolicyConfigurationManager());
+        }
+    }
+
+    @Override
+    public boolean isAuthorized(String applicationName, String moduleName, String beanName, String methodName, String methodInterface,
+                                String methodSignature, List<Object> methodParameters, EnterpriseBean bean, Subject subject) {
+        JaccService jaccService = jaccServiceRef.getService();
+        if (jaccService != null) {
+            String[] methodSignatureArray = convertMethodSignature(methodSignature);
+            final EJBMethodPermission ejbPerm = new EJBMethodPermission(beanName, methodName, methodInterface, methodSignatureArray);
+            return checkResourceConstraints(jaccService.getContextId(applicationName, moduleName), methodParameters, bean, ejbPerm, subject, jaccService.getPolicyProxy());
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isSubjectInRole(String applicationName, String moduleName, String beanName, String methodName,
+                                   List<Object> methodParameters, String role, EnterpriseBean bean, Subject subject) {
+
+        JaccService jaccService = jaccServiceRef.getService();
+        if (jaccService != null) {
+            final EJBRoleRefPermission ejbPerm = new EJBRoleRefPermission(beanName, role);
+            return checkResourceConstraints(jaccService.getContextId(applicationName, moduleName), methodParameters, bean, ejbPerm, subject, jaccService.getPolicyProxy());
+        }
+        return false;
+    }
+
+    private boolean checkResourceConstraints(String contextId, List<Object> methodParameters, EnterpriseBean bean, Permission ejbPerm, Subject subject, PolicyProxy policyProxy) {
+        boolean result = false;
+        final Object[] ma = null;
+
+        /*
+         * TODO Doesn't seem to handle EJB-3.0 annotated beans.
+         */
+        if (methodParameters != null && methodParameters.size() > 0) {
+            methodParameters.toArray(new Object[methodParameters.size()]);
+        }
```

Comment:
What purpose is this serving if the result isn't captured? I see this is just part of what was taken from EJBSecurityValidatorImpl.java, but might consider taking this out unless there's some justification you can think of.

Reply:
I believe that `ma` should be assigned to the return value of toArray and that it is a bug that it isn't today.  Yes it is the same as it was before and I didn't change it since it would change behavior.   Below is what I think that it should be changed to.  Please let me know if you agree.

```suggestion
        /*
         * TODO Doesn't seem to handle EJB-3.0 annotated beans.
         */
        if (methodParameters != null && methodParameters.size() > 0) {
            ma = methodParameters.toArray(new Object[methodParameters.size()]);
        }
```

Reply:
I agree that that seems like the right change to make.